src/components/stream-source/component/StreamGenerator.ts

```diff
@@ -0,0 +1,77 @@
+interface DeferredPrimise {
+  promise?: Promise<unknown>;
+  resolve?(value?: unknown): void;
+  reject?(): void;
+}
+
+enum StreamStatus {
+  ACTIVE,
+  CLOSED,
+}
+
+class StreamGenerator {
+  private _next: DeferredPrimise = StreamGenerator.createDeferredPromise();
+  private _status: StreamStatus = StreamStatus.ACTIVE;
+  private _buffer: string[] = [];
+  private _err?: Error;
+  private _stream: AsyncGenerator<string>;
+  private _loadingSignal: DeferredPrimise =
+    StreamGenerator.createDeferredPromise();
+
+  private static createDeferredPromise() {
+    const deferred: DeferredPrimise = {};
+    deferred.promise = new Promise((_resolve, _reject) => {
+      deferred.resolve = _resolve;
+      deferred.reject = _reject;
+    });
+    return deferred;
+  }
+
+  private async *generator() {
+    while (this.status === StreamStatus.ACTIVE) {
```

Comment:
I think this will cause a race condition, since the producer of the stream can call `done` before the consumer has finished processing it.

Reply:
Can you explain more details?